### ISSUE TITLE
SHM provider fails init when stale shm files are present

### ISSUE THREAD

**Issue by rwespetal** [ofiwg/libfabric#6476]
We're seeing an issue with the SHM provider initialization when stale shared memory segment files are present:

```
libfabric:62787:shm:ep_ctrl:smr_create():164<warn> shm_open error
libfabric:62787:shm:ep_ctrl:smr_create():164<warn> shm_open error
Process 0 sending 10 to 1, tag 201 (32 processes in ring)
[ip-172-31-13-34:06020] [[40439,1],1] selected pml cm, but peer [[40439,1],0] on ip-172-31-3-231 selected pml ob1
[ip-172-31-13-34:06020] mtl_ofi_endpoint.h:49: *** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
```

I confirmed that it's returning `-EEXIST` and cleaning `/dev/shm` resolves this so I believe it's due to PR #6468. Is there anything we can do to safely detect this and clean up when the SHM signal handler to cleanup these files is not run?

**Comment by aingerson** [comment]
@rwespetal Hmm, why are the signal handlers not getting called? Why are these regions not getting cleaned up?

**Comment by rwespetal** [comment]
I believe the signal handlers will not be called when SIGKILL is sent. I noticed this after Libfabric crashed.

**Comment by wzamazon** [comment]
Yes, for example when the system ran out of memory process will be killed by OOM by SIGKILL.

**Comment by aingerson** [comment]
I'm not sure if there's a clean way to handle this in libfabric. It seems to me that either forcing SIGKILL to a process manually or that sent by OOM because of a lack of memory is more of a system issue that needs to be resolved outside of libfabric. Since there's no way to catch/cleanup after a SIGKILL, I think this has to be done outside of libfabric.
The only other thing I can think of is if we remove the name solution with node/service and disable the user setname function so that the EP name is always the shm generated name (ie PID:UID:dom_idx:ep_idx) and then we can revert the change that caused this, allowing the EP to force reopen the shm region even if it already exists.

**Comment by wzamazon** [comment]
@aingerson Would you please elaborate on the problem the original commit try to fix? I feel like I do not have the full context.

**Comment by aingerson** [comment]
The original problem was that two EPs were opening regions with the same name (using the node/service format). The second one, instead of detecting that it was a duplicate, would simply recreate the first peer's region and there were segfaults/communication issues when both peers were using the same region as its own. I added the O_EXCL flag so that the shm_open would fail if the region already existed to allow the peer to either set a new name or fail cleanly instead of trying to use the same region for two peers.
If we removed the ability to force a specific name (like with the node/service) then we can remove the O_EXCL flag since all region names will be unique to the PID/UID so it's safe to override it if it already exists.

**Comment by wzamazon** [comment]
Does the shm ep name have to be same as shm file name? I think the ability to force a specific EP name is part of the API, so we might not want to disable it. Meanwhile, the shm file name is an implementation detail, which does not need to be same as EP name. Specifically, I think we can use the user forced name as a prefix, append it with pid, etc and make shm file name.

**Comment by aingerson** [comment]
If we change the name by appending the PID for example, it becomes difficult for peers to connect to each other since they somehow need to have the full name/PID in order to open the region and and send a message. When the peers exchange names, they call getname() and send that name so we would have to return the full shm name there in order for the peer to be able to open the correct region.

**Comment by rwespetal** [comment]
I see. Perhaps we can only pass O_EXCL when the application uses node/service and allow the file to be reused if it is the generated name?

**Comment by aingerson** [comment]
Potentially. But it seems like some very specific extra code to avoid something that would still be possible in the node/service case. Like if a run using the node/service/setname was killed by the OOM using SIGKILL, the next run would still fail. In that case, there would still need to be some sort of cleanup after the SIGKILL.

**Comment by rwespetal** [comment]
Right, I'd prefer that we can recover from SIGKILL in general given it will be difficult for users to know what happened and how to fix it.

I'm not sure I understand why it will be more difficult to connect if a unique address is used. getname() is not valid until the endpoint is enabled. Or is there a mechanism that allows peers to know the shm addresses implicitly without an address exchange?

**Comment by aingerson** [comment]
I think it would prevent the ability to exchange the names in band. In order to send a message to a peer, the sender needs to know the name of the region in order to connect to it and send it's own name. So at least one side needs to have a known name of some sort.
There is some code for a name server which sort of acts as an internal out of band exchange. We might be able to look into using that or a similar method?

**Comment by rwespetal** [comment]
Oh, makes sense. Exchanging addresses or shm file names internally sounds like it would work.

**Comment by shefty** [comment]
Exchanging names OOB still requires some sort of well-known address.  Multiple jobs running on the same system would be problematic.

**Comment by wzamazon** [comment]
@aingerson

If I understand the issue correctly, open shm endpoint with node/server will never work. What the O_EXCL patch does is when node/service is used, we  return error instead of segfault.

If this is right, I think what you suggested. e.g. "removed the ability to force a specific name (like with the node/service) " is a reasonable approach, because force a name with node/service would not work.

**Comment by shefty** [comment]
The use of a specific address makes it easy for a job cleanup manager to remove stale files.

**Comment by aingerson** [comment]
@wzamazon Not exactly. The node/service should work. The case where it doesn't work right now is when the node/service is used and then the process is killed with SIGKILL, causing the next run with that name to fail. Before the EXCL patch, what was happening is that the two EPs could have the same name (node/service) and cause segfaults because they are using the same region. Essentially, the patch is just making sure that two EPs can't use the same name. 
While removing the ability to force a name could potentially avoid the issue, I think a similar problem still remains which is that there are still stale files left over after the SIGKILL. The approach I suggested is not going to help remove those files, just avoid failing the next time those are used. In addition, it's going to introduce more issues like the exchange of names and the fact that we're removing support for an API call which I don't like.
I think the ideal solution for this problem is for the OOM or whatever daemon/person sent the SIGKILL to be responsible for cleaning up the files. There's no way for libfabric to catch the signal and doing nothing about it could leave a bunch of stale files around. And as Sean pointed out, if you opt for known addresses, then a cleanup manager could easily do this for you.

**Comment by rwespetal** [comment]
I'm not sure I agree although I understand the concerns. The problem is the current failure mode gives the user no indication on what happened or how to fix it, it also requires the user to take a manual action which isn't ideal. SIGKILL will happen to users whether it is OOM, crash, or issued by whatever started the app. We also don't have hooks into the job schedulers to clean up like some of the MPI implementations do.

**Comment by shefty** [comment]
The O_EXCL flag causes open to fail if the filename already exists.  If the process that opened the file is still running, the failure is desired.  If the process is not running, ideally, the open would be retried and the file replaced.

Sooo.... can we determine if the process that opened the file is still running?

One option that might work is to require that all shm names be strings of udp/ip addresses.  Before opening a file, the shm provider can open and bind a udp socket.  If that succeeds, then it can open the shm file (O_EXCL should not be needed).  If socket bind fails, then there is some other active process using the file, and we can abort.

If the application does not call fi_setname(), we can bind to a wildcard port and use whatever was returned.

**Comment by wzamazon** [comment]
@aingerson

I guess I do not understand how node/service could work with shm.

Earlier, you mentioned that:

```
The node/service should work. The case where it doesn't work right now is when the node/service is used and then the process is killed with SIGKILL, causing the next run with that name to fail.
```
This is puzzling to me. I thought node/service in general does not work with shm, because all processes try to use same shm file name. This is independent from the SIGKILL issue. However, it seems node/service can work. I wonder how?

**Comment by rwespetal** [comment]
Sean, do you consider this a blocker for 1.12? I'm concerned that EFA users are going to be impacted by this given how often we see it in various internal CI runs.

**Comment by shefty** [comment]
I would like to have a fix prior to the 1.12 release.

**Comment by aingerson** [comment]
See #6497 I wrote a quick patch using the PID solution just to get something working at least for now. It works for me when I manually kill and restart a process. Let me know if you still want me to work on the socket method if that would be safer.
